### PR TITLE
Add release test runnning full asan python test

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
This PR adds a ray project that run the full python test suite with asan build enabled.

With the session started, run "setup", and then "run". If you need to checkout a new branch, run "recompile" with "git_sha" and "run". 

Current output
```
//python/ray/tests:test_actor                                            PASSED in 134.2s
//python/ray/tests:test_actor_advanced                                   PASSED in 132.5s
//python/ray/tests:test_actor_failures                                   PASSED in 91.3s
//python/ray/tests:test_actor_pool                                       PASSED in 20.6s
//python/ray/tests:test_actor_resources                                  PASSED in 93.5s
//python/ray/tests:test_advanced                                         PASSED in 69.9s
//python/ray/tests:test_advanced_2                                       PASSED in 104.2s
//python/ray/tests:test_advanced_3                                       PASSED in 103.4s
//python/ray/tests:test_args                                             PASSED in 17.3s
//python/ray/tests:test_array                                            PASSED in 32.7s
//python/ray/tests:test_asyncio                                          PASSED in 16.0s
//python/ray/tests:test_autoscaler                                       PASSED in 5.3s
//python/ray/tests:test_autoscaler_aws                                   PASSED in 1.3s
//python/ray/tests:test_autoscaler_yaml                                  PASSED in 2.2s
//python/ray/tests:test_basic_2                                          PASSED in 131.4s
//python/ray/tests:test_component_failures                               PASSED in 21.0s
//python/ray/tests:test_component_failures_2                             PASSED in 44.9s
//python/ray/tests:test_component_failures_3                             PASSED in 14.6s
  WARNING: //python/ray/tests:test_component_failures_3: Test execution time (14.6s excluding execution overhead) outside of range for MODERATE tests. Consider setting timeout="short" or size="small".
//python/ray/tests:test_dynres                                           PASSED in 68.9s
//python/ray/tests:test_failure                                          PASSED in 168.5s
//python/ray/tests:test_global_gc                                        PASSED in 34.5s
//python/ray/tests:test_global_state                                     PASSED in 20.0s
//python/ray/tests:test_iter                                             PASSED in 108.8s
//python/ray/tests:test_joblib                                           PASSED in 63.8s
//python/ray/tests:test_memory_limits                                    PASSED in 90.2s
//python/ray/tests:test_memstat                                          PASSED in 18.2s
//python/ray/tests:test_metrics_export                                   PASSED in 1.7s
//python/ray/tests:test_microbenchmarks                                  PASSED in 17.4s
//python/ray/tests:test_mini                                             PASSED in 0.8s
//python/ray/tests:test_multi_node                                       PASSED in 206.3s
//python/ray/tests:test_multi_node_2                                     PASSED in 30.8s
//python/ray/tests:test_multinode_failures                               PASSED in 84.8s
//python/ray/tests:test_multinode_failures_2                             PASSED in 17.1s
  WARNING: //python/ray/tests:test_multinode_failures_2: Test execution time (17.1s excluding execution overhead) outside of range for MODERATE tests. Consider setting timeout="short" or size="small".
//python/ray/tests:test_multiprocessing                                  PASSED in 77.6s
//python/ray/tests:test_node_manager                                     PASSED in 7.4s
//python/ray/tests:test_numba                                            PASSED in 5.6s
//python/ray/tests:test_object_manager                                   PASSED in 36.6s
//python/ray/tests:test_projects                                         PASSED in 3.2s
//python/ray/tests:test_queue                                            PASSED in 12.1s
//python/ray/tests:test_ray_init                                         PASSED in 10.6s
//python/ray/tests:test_reconstruction                                   PASSED in 54.9s
//python/ray/tests:test_reference_counting                               PASSED in 149.4s
//python/ray/tests:test_reference_counting_2                             PASSED in 151.7s
//python/ray/tests:test_serialization                                    PASSED in 35.7s
//python/ray/tests:test_tempfile                                         PASSED in 36.9s
//python/ray/tests:test_tensorflow                                       PASSED in 41.0s
//python/ray/tests:test_webui                                            PASSED in 4.3s
//python/ray/tests:test_basic                                             FLAKY, failed in 1 out of 2 in 97.0s
  Stats over 2 runs: max = 97.0s, min = 20.7s, avg = 58.8s, dev = 38.1s
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_basic/test_attempts/attempt_1.log
//python/ray/tests:test_cancel                                            FLAKY, failed in 2 out of 3 in 315.0s
  Stats over 3 runs: max = 315.0s, min = 94.2s, avg = 241.4s, dev = 104.1s
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_cancel/test_attempts/attempt_1.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_cancel/test_attempts/attempt_2.log
//python/ray/tests:test_cython                                           FAILED in 3 out of 3 in 0.8s
  Stats over 3 runs: max = 0.8s, min = 0.8s, avg = 0.8s, dev = 0.0s
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_cython/test.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_cython/test_attempts/attempt_1.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_cython/test_attempts/attempt_2.log
//python/ray/tests:test_debug_tools                                      FAILED in 3 out of 3 in 11.7s
  Stats over 3 runs: max = 11.7s, min = 11.7s, avg = 11.7s, dev = 0.0s
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_debug_tools/test.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_debug_tools/test_attempts/attempt_1.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_debug_tools/test_attempts/attempt_2.log
//python/ray/tests:test_memory_scheduling                                FAILED in 3 out of 3 in 114.6s
  Stats over 3 runs: max = 114.6s, min = 111.4s, avg = 113.0s, dev = 1.3s
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_memory_scheduling/test.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_memory_scheduling/test_attempts/attempt_1.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_memory_scheduling/test_attempts/attempt_2.log
//python/ray/tests:test_metrics                                          FAILED in 3 out of 3 in 37.7s
  Stats over 3 runs: max = 37.7s, min = 34.7s, avg = 35.9s, dev = 1.3s
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_metrics/test.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_metrics/test_attempts/attempt_1.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_metrics/test_attempts/attempt_2.log
//python/ray/tests:test_stress                                           FAILED in 3 out of 3 in 6.5s
  Stats over 3 runs: max = 6.5s, min = 6.4s, avg = 6.4s, dev = 0.0s
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_stress/test.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_stress/test_attempts/attempt_1.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_stress/test_attempts/attempt_2.log
//python/ray/tests:test_stress_sharded                                   FAILED in 3 out of 3 in 8.7s
  Stats over 3 runs: max = 8.7s, min = 8.6s, avg = 8.7s, dev = 0.0s
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_stress_sharded/test.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_stress_sharded/test_attempts/attempt_1.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_stress_sharded/test_attempts/attempt_2.log
//python/ray/tests:test_unreconstructable_errors                         FAILED in 3 out of 3 in 2.0s
  Stats over 3 runs: max = 2.0s, min = 2.0s, avg = 2.0s, dev = 0.0s
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_unreconstructable_errors/test.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_unreconstructable_errors/test_attempts/attempt_1.log
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/2139a9ac91f0a6877cfd4920666e807a/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_unreconstructable_errors/test_attempts/attempt_2.log
```

ASAN didn't catch anything specific. Many of the failed tests come from failing to start redis, even though this machine has 64Gb memory:
```
2020-06-08 01:35:56,086 INFO resource_spec.py:212 -- Starting Ray with 57.96 GiB memory available for workers and up to 0.47 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
2020-06-08 01:35:56,204 WARNING services.py:923 -- Redis failed to start, retrying now.
2020-06-08 01:35:56,454 WARNING services.py:923 -- Redis failed to start, retrying now.
________________ ERROR at setup of test_getting_many_objects[4] ________________

request = <SubRequest 'ray_start_sharded' for <Function test_getting_many_objects[4]>>

    @pytest.fixture(params=[1, 4])
    def ray_start_sharded(request):
        num_redis_shards = request.param
    
        if os.environ.get("RAY_USE_NEW_GCS") == "on":
            num_redis_shards = 1
            # For now, RAY_USE_NEW_GCS supports 1 shard, and credis supports
            # 1-node chain for that shard only.
    
        # Start the Ray processes.
        ray.init(
            object_store_memory=int(0.5 * 10**9),
            num_cpus=10,
            num_redis_shards=num_redis_shards,
>           redis_max_memory=10**7)

/home/ubuntu/ray/python/ray/tests/test_stress_sharded.py:22: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/ubuntu/ray/python/ray/worker.py:723: in init
    ray_params=ray_params)
/home/ubuntu/ray/python/ray/node.py:180: in __init__
    self.start_head_processes()
/home/ubuntu/ray/python/ray/node.py:681: in start_head_processes
    self.start_redis()
/home/ubuntu/ray/python/ray/node.py:502: in start_redis
    fate_share=self.kernel_fate_share)
/home/ubuntu/ray/python/ray/services.py:816: in start_redis
    fate_share=fate_share)
/home/ubuntu/ray/python/ray/services.py:1006: in _start_redis_instance
    redis_client.set("redis_start_time", time.time())
/home/ubuntu/anaconda3/lib/python3.6/site-packages/redis/client.py:1766: in set
    return self.execute_command('SET', *pieces)
/home/ubuntu/anaconda3/lib/python3.6/site-packages/redis/client.py:878: in execute_command
    return self.parse_response(conn, command_name, **options)
/home/ubuntu/anaconda3/lib/python3.6/site-packages/redis/client.py:892: in parse_response
    response = connection.read_response()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = Connection<host=127.0.0.1,port=39641,db=0>

    def read_response(self):
        "Read the response from a previously sent command"
        try:
            response = self._parser.read_response()
        except socket.timeout:
            self.disconnect()
            raise TimeoutError("Timeout reading from %s:%s" %
                               (self.host, self.port))
        except socket.error:
            self.disconnect()
            e = sys.exc_info()[1]
            raise ConnectionError("Error while reading from %s:%s : %s" %
                                  (self.host, self.port, e.args))
        except:  # noqa: E722
            self.disconnect()
            raise
    
        if self.health_check_interval:
            self.next_health_check = time() + self.health_check_interval
    
        if isinstance(response, ResponseError):
>           raise response
E           redis.exceptions.ResponseError: OOM command not allowed when used memory > 'maxmemory'.
```

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: ci/asan_tests/ray-project/cluster.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# This file is generated by `ray project create`.
+
+# A unique identifier for the head node and workers of this cluster.
+cluster_name: asan_python_tests
+
+# The maximum number of workers nodes to launch in addition to the head
+# node. This takes precedence over min_workers. min_workers defaults to 0.
+max_workers: 1
```
Comment (rkooo567) [inline ci/asan_tests/ray-project/cluster.yaml:8]: Why not having deterministic number of workers? Is it intended?

File: ci/asan_tests/ray-project/cluster.yaml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    - cd $HOME; rm -rf $HOME/ray; git clone https://github.com/ray-project/ray || true
+    # Install Node.js in order to build the dashboard.
+    - curl -sL https://deb.nodesource.com/setup_12.x | sudo -E bash
+    - sudo apt-get install -y nodejs || true
+    - cd ray/python/ray/dashboard/client; npm ci; npm run build
+    # Set up Ray dependencies.
+    - which bazel || bash -c "cd $HOME/ray; ci/travis/install-bazel.sh"
+
+# Command to start ray on the head node. You don't need to change this.
+head_start_ray_commands: []
```
Comment (rkooo567) [inline ci/asan_tests/ray-project/cluster.yaml:49]: Newline

File: ci/asan_tests/ray-project/project.yaml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+# Pathnames for files and directories that should be saved
+# in a snapshot but that should not be synced with a# session. Pathnames can be relative to the project
+# directory or absolute. Generally, this should be files
+# that were created by an active session, such as
+# application checkpoints and logs.
+output_files: [
+  # For example, uncomment this to save the logs from the
+  # last ray job.
+  # "/tmp/ray/session_latest",
+]
```
Comment (rkooo567) [inline ci/asan_tests/ray-project/project.yaml:38]: Newline

File: ci/asan_tests/run_asan_tests.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+export LD_PRELOAD=/usr/lib/gcc/x86_64-linux-gnu/7/libasan.so
+export ASAN_OPTIONS=detect_leaks=0
+
+cd $HOME/ray
+# async plasma test
+python -m pytest -v --durations=5 --timeout=300 python/ray/experimental/test/async_test.py
+
+# Ray tests
+bazel test --test_tag_filters=-jenkins_only python/ray/serve/...
+bazel test --test_tag_filters=-jenkins_only python/ray/dashboard/...
```
Comment (rkooo567) [inline ci/asan_tests/run_asan_tests.sh:13]: Do we have the dashboard tests?

File: ci/asan_tests/bazelrc [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# Cache
```
Comment (rkooo567) [inline ci/asan_tests/bazelrc:1]: If `root/.bazelrc` changes ASAN options, this must be also changed right? Would you mind adding comments at both files to explain that?

**Review by rkooo567 — APPROVED** [review]
Thanks a lot!! This will be really helpful! We should also think about process to react to ASAN test failures I guess.

**Comment by mehrdadn** [comment]
A bit related: I cleaned up a lot of Bazel config options in #8828 so that we can have most of them in `.bazelrc` (instead of scattered across shell scripts etc.). Let me know if we should coordinate anything with this PR.

File: ci/asan_tests/Makefile [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+	cd ${HOME}/ray/doc/examples/cython; python setup.py install --user
+
+	echo "Settting up the shell"
+	cp bazelrc ${HOME}/.bazelrc # Setup cache
+	echo "LD_PRELOAD=/usr/lib/gcc/x86_64-linux-gnu/7/libasan.so" >> ${HOME}/.bashrc
+	echo "ASAN_OPTIONS=detect_leaks=0" >> ${HOME}/.bashrc
+
+	echo "Compiling ray"
+	cd ${HOME}/ray; git fetch; git pull origin master
+	cd ${HOME}/ray/python; pip install -e . --verbose || true
```
Comment (mehrdadn) [inline ci/asan_tests/Makefile:21]: Add quotes around variables (`${HOME}` etc.)

File: ci/asan_tests/run_asan_tests.sh (hunk 4 above)
Comment (simon-mo) [reply ci/asan_tests/run_asan_tests.sh:13]: yes there's a dashboard unit test @mfitton added

File: ci/asan_tests/ray-project/cluster.yaml (hunk 1 above)
Comment (simon-mo) [reply ci/asan_tests/ray-project/cluster.yaml:8]: nope. this is just the default generated by `ray project init`

File: ci/asan_tests/bazelrc [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+build --remote_cache=https://storage.googleapis.com/ray-bazel-cache
+build --disk_cache=/tmp/bazel-cache
+build --repository_cache=/tmp/bazel-cache
+build --remote_upload_local_results=false
+
+# Asan
+build --strip=never
```
Comment (mehrdadn) [inline ci/asan_tests/bazelrc:8]: Do we need separate config files for these? We already have `build:asan` in the root `.bazelrc` -- do these work better than `bazel --config`?

**Review by edoakes — APPROVED** [review]
Looks great! @rkooo567 let's start running these with the upcoming release

File: ci/asan_tests/Makefile [hunk 8]
```diff
@@ -0,0 +1,28 @@
+SHELL=/bin/bash
```
Comment (mehrdadn) [inline ci/asan_tests/Makefile:1]: We shouldn't use a Makefile for this by the way; it should be a script of some sort. (It also adds an unnecessary Make dependency.) We can keep it for now if it's time-sensitive, but let's turn it into a script later.

File: ci/asan_tests/bazelrc (hunk 7 above)
Comment (simon-mo) [reply ci/asan_tests/bazelrc:8]: Added a reason. We don't want to use the --config=. In both the python wheel building step and the testing step.

**Pull request merged**